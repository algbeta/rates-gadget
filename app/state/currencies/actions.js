import API_URLS from '../../utils/apiUrls';

const actionTypes = {
  FETCH_CURRENCIES: 'FETCH_CURRENCIES',
  FETCH_CURRENCIES_SUCCESS: 'FETCH_CURRENCIES_SUCCESS',
  FETCH_CURRENCIES_FAIL: 'FETCH_CURRENCIES_FAIL',
};

const fetchCurrencies = () => ({
  type: actionTypes.FETCH_CURRENCIES,
  url: API_URLS.getLatest(),
});

const fetchCurrenciesFail = error => ({
  type: actionTypes.FETCH_CURRENCIES_FAIL,
  error,
});

const fetchCurrenciesSuccess = data => ({
  type: actionTypes.FETCH_CURRENCIES_SUCCESS,
  data,
});

export {
  actionTypes,
  fetchCurrencies,
  fetchCurrenciesFail,
  fetchCurrenciesSuccess,
};
