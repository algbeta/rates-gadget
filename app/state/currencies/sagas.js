import { put, takeLatest } from 'redux-saga/effects';
import { apiGet } from '../common/api';
import API_URLS from '../../utils/apiUrls';
import {
  fetchCurrenciesSuccess,
  fetchCurrenciesFail
} from './actions';

export function* fetchCurrencies() {
  yield takeLatest(actionTypes.FETCH_CURRENCIES, function*() {
  let response = {};
  try {
    response = yield apiGet(API_URLS.getLatest());
  } catch (e) {
    yield put(fetchCurrenciesFail(e));
    return;
  }

  yield put(fetchCurrenciesSuccess(response));
  });
}
