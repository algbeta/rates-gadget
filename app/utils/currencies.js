const currencies = {
  EUR: {
    name: 'Euro',
    shortcut: 'EUR',
    sign: '€',
  },
  GBP: {
    name: 'British Pound',
    shortcut: 'GBP',
    sign: '£',
  },
  USD: {
    name: 'American Dollar',
    shortcut: 'USD',
    sign: '$',
  },
};

const currencyTypeToSelect = {
  from: 'from',
  to: 'to',
};

const forceTwoDigitsAfterDot = number => +number.toFixed(2);

export { currencyTypeToSelect, forceTwoDigitsAfterDot };
export default currencies;
