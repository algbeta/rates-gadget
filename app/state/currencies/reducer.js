import { actionTypes } from './actions';
import currencies, { currencyTypeToSelect } from '../../utils/currencies';
const initialState = {
  amount: 0,
  [currencyTypeToSelect.from]: currencies.GBP.shortcut,
  [currencyTypeToSelect.to]: currencies.USD.shortcut,
};

export default function(state = initialState, action) {
  switch (action.type) {
    case actionTypes.SET_EXCHANGE_AMOUNT: {
      return {
        ...state,
        amount: action.amount,
      };
    }
    case actionTypes.FETCH_CURRENCIES_SUCCESS: {
      return {
        ...state,
        [action.base]: {
          ...action.rates,
        },
      };
    }
    default:
      return state;
  }
}
