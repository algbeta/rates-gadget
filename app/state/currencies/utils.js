import currencies from '../../utils/currencies';

function processResponse(base, rates) {
  const processedResponse = {};
  const currenciesArray = Object.keys(currencies);
  processedResponse[base] = rates;

  Object.keys(rates).forEach(cur => {
    processedResponse[cur] = {};
    processedResponse[cur][base] = 1 / rates[cur];
  });

  currenciesArray.forEach(cur => {
    currenciesArray.forEach(propCur => {
      if (propCur === cur) {
        processedResponse[cur][propCur] = 1;
      } else if (processedResponse[cur] && !processedResponse[cur][propCur]) {
        processedResponse[cur][propCur] =
          processedResponse[base][propCur] / processedResponse[base][cur];
      }
    });
  });

  return processedResponse;
}

export { processResponse };
