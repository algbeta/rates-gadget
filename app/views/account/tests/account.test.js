import React from 'react';
import renderer from 'react-test-renderer';
import Account from '../account';
import currencies from '../../../utils/currencies';

test('Account renders correctly', () => {
  const tree = renderer
    .create(<Account currency={currencies.USD.shortcut} value={1500} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
