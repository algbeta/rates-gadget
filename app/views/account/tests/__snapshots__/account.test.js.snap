// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account renders correctly 1`] = `
<div>
  <span>
     Available funds: 
  </span>
  <span>
    1500
     
    $
  </span>
</div>
`;
